src/core/audio/AudioAlsa.cpp

```diff
@@ -139,6 +142,59 @@ QString AudioAlsa::probeDevice()
 
 
 
+/**
+ * @brief Creates a list of all available devices.
+ *
+ * Uses the hints API of ALSA to collect all devices. This also includes plug
+ * devices. The reason to collect these and not the raw hardware devices
+ * (e.g. hw:0,0) is that hardware devices often have a very limited number of
+ * supported formats, etc. Plugs on the other hand are software components that
+ * map all types of formats and inputs to the hardware and therefore they are
+ * much more flexible and more what we want.
+ *
+ * Further helpful info http://jan.newmarch.name/LinuxSound/Sampled/Alsa/.
+ *
+ * @return A collection of devices found on the system.
+ */
+AudioAlsa::DeviceInfoCollection AudioAlsa::getAvailableDevices()
+{
+	DeviceInfoCollection deviceInfos;
+
+	char **hints;
+
+	/* Enumerate sound devices */
+	int err = snd_device_name_hint(-1, "pcm", (void***)&hints);
+	if (err != 0)
+	{
+		return deviceInfos;
+	}
+
+	char** n = hints;
+	while (*n != NULL)
+	{
+		char *name = snd_device_name_get_hint(*n, "NAME");
+		char *description = snd_device_name_get_hint(*n, "DESC");
+
+		if (name != 0 && description != 0)
+		{
+			deviceInfos.push_back(DeviceInfo(QString(name), QString(description)));
+		}
+
+		free(name);
+		free(description);
+
```

Comment:
I remember seeing this earlier & looking into it. Turns out it's perfectly valid to free a null pointer in C/C++: http://www.cplusplus.com/reference/cstdlib/free/

Additionally, moving BOTH calls to `free` into the if statement would cause a memory leak in the case that only one of them is null. So I vote to leave as-is.